packages/@lwc/engine-core/src/framework/api.ts

```diff
@@ -204,7 +204,11 @@ const ElementHook: Hooks<VElement> = {
             const { props } = vnode.data;
             if (!isUndefined(props) && !isUndefined(props.innerHTML)) {
                 if (elm.innerHTML === props.innerHTML) {
-                    delete props.innerHTML;
+                    // Do a shallow clone since VNodeData may be shared across VNodes due to hoist optimization
+                    vnode.data = {
+                        ...vnode.data,
+                        props: cloneAndOmitKey(props, 'innerHTML'),
+                    };
```

Comment:
When reaching this code path, there is no way the `vnode.data` is hoisted as the `innerHTML` prop requires an invocation to `api_sanitize_html_content`. It is safe to mutate it in place by case it to `any`.

By the way, this block of code looks a little misplaced.

Reply:
I would kind of prefer to leave it as a clone. It seems unsafe for a developer to rely on quirks of the optimization algorithm. In the future, we may very well want to hoist:

```js
api_sanitize_html_content("static string")
```

In fact, we are already caching the output on the `$ctx` anyway:

https://github.com/salesforce/lwc/blob/52e0a2b3a0463df42baa161cfd369aa069d96303/packages/%40lwc/template-compiler/src/__tests__/fixtures/directive-inner-html/valid/expected.js#L8-L14

So if we apply the hosting optimization, arguably we may not need the `$ctx` optimization anymore.

Reply:
uff, uff! `elm.innerHTML === props.innerHTML` broke my heart! I will have never approve that change. The whole point of the diffing algo is to avoid touching the DOM at all cost to compare things... reading the innerHTML from the element to compare it with the vnode is just wrong. We should fix that.

Additionally, the innerHTML should NOT be a regular property, it should have been a snabbdom module who's job is to compare two vnodes to determine whether or not the innerHTML needs to be set, rather than tricking the props module to not see the innerHTML for some cases when we know that it is the same. This is obviously not related to this PR, but needs to be fixed at some point. I might be missing something obvious here.

Reply:
@caridy Opened an issue to track this: https://github.com/salesforce/lwc/issues/2603